book/visual-effects.md

```diff
@@ -982,56 +903,49 @@ color mixing.
-(e.g. ones created via [`z-index`][z-index] do not). However, for ease of
-implementation, all visual effects in CSS that generally require surfaces to
-implement are specified to go hand-in-hand with a stacking context, so the tree
-of stacking contexts is very related to the tree of surfaces.
+``` {.python}
+def parse_color(color):
+    # ...
+    elif color.startswith("#") and len(color) == 9:
+        r = int(color[1:3], 16)
+        g = int(color[3:5], 16)
+        b = int(color[5:7], 16)
+        a = int(color[7:9], 16)
+        return skia.Color(r, g, b, a)
+    # ...
+```
 
-[stacking-context]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context
-[z-index]: https://developer.mozilla.org/en-US/docs/Web/CSS/z-index
+Check that your browser renders dark-orange text for the example
+above. That shows that it is actually mixing the black color with the
+existing orange color from the background.
 
-To match this use pattern, in Skia, surfaces form a stack. You can
-push a new surface on the stack, raster things to it, and then pop it
-off by blending it with surface below. When traversing the tree of
-stacking contexts, you push a new surface onto the stack every time
-you recurse into a new stacking context, and pop-and-blend every time
-you return from a child stacking context to its parent.
- 
-In real browsers, stacking contexts are formed by HTML elements with
-certain styles, up to any descendants that themselves have such
-styles. The full definition is actually quite complicated, so in this
-chapter we'll simplify by treating every layout object as a stacking
-context.
+However, there's another, subtle different way to create transparency
+with CSS. Here, 50% transparency is applied to the whole element using
+the `opacity` property:
+
+<div style="font-size: 50px; padding: 15px; text-align: center;
+    background: orange; color: black; opacity: .5">Test</div>
+    
+Now the opacity applies to both the background and the text, so the
+background is now a little lighter. But note that the text
+is now gray, not dark orange. The black and orange pixels are no
+longer blended together!
+
+That's because opacity introduces what CSS calls a [stacking
+context][stacking-context]. The details aren't important, but the
+order of operations is. In the first example, the black pixels were
+first made transparent, then blended with the background. Thus, 50%
+transparent black pixels were blending with orange pixels, resulting
+in a dark-orange color. In the second example, the black pixels were
+first blended with the background, then the result was made
+transparent. Thus, fully black pixels replaced fully orange ones,
+resulting just black pixels, which were later made 50% transparent.
+
+Applying effect in the proper order, as necessary to implement effects
+like `opacity`, requires more careful handling of surfaces.
```

Comment:
Maybe link to this blog post on the subject for more fun details?

https://dbaron.org/log/20130306-compositing-blending